scripts/rpc-maas-tool.py

```diff
@@ -292,13 +309,199 @@ def main(self):
+        for check in self.rpmac.checks.values():
+            for alarm in check.alarms:
+                try:
+                    self._parse_alarm_criteria(alarm)
+                except ValueError as e:
+                    LOGGER.info(e)
+                    rc = 1
+        return rc
+
+    def _find_metrics(self, ast, metrics):
+        """Recursively descend AST looking for metricName elements
+
+        When a metric name is found, a string is constructed from all it's
+        single character child nodes, a list of metric name strings is
+        returned.
+        """
+        if hasattr(ast, 'type') and ast.type == 'metricName':
+            name_node = ast.children[0]
+            name_str = ''.join(map(str, name_node.children))
+            metrics.append(name_str)
+        if hasattr(ast, 'children'):
+            for child in ast.children:
+                child_metrics = self._find_metrics(child, [])
+                if child_metrics:
+                    metrics.extend(child_metrics)
+        return metrics
+
+    def verify_local(self):
+        """Checks MaaS configuration without using MaaS API
+
+        Checks three things:
+        1) Execute the command defined in each check and check its return code
+        2) Compile the alarm criteria and check for syntax
+        3) Check the metrics required by the alarm criteria against the metrics
+            produced by executing the check commands.
+        """
+
+        status_line_re = re.compile('status\s+(?P<status>.*)')
+        metric_line_re = re.compile(
+            'metric\s+(?P<name>[^\s]+)\s+(?P<unit>[^\s]+)\s+(?P<value>[^\s]+)')
+
+        def _alarm_metrics_from_check(check):
+            """Get all the metrics referenced by a check's alarms"""
+            metrics = []
+            for alarm in check['alarms'].values():
+                ast = self._parse_alarm_criteria(alarm)
+                return self._find_metrics(ast, metrics)
+
+        def _execute_check(args, check, rpcm):
+            """Execute one check
+
+            This function will be called from a threadpool thread.
+            """
+            try:
+                result = {'success': True,
+                          'output': subprocess.check_output(
+                              args,
+                              stderr=subprocess.STDOUT),
+                          'check': check
+                          }
+            except subprocess.CalledProcessError as e:
+                result = {'success': False,
+                          'output': e.output,
+                          'check': check
+                          }
+            rpcm.q.put(result)
+
+        # Checks are executed using a threadpool to speed up verification.
+        execution_results = []
+        with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
+            for check in self.rpcmac.checks.values():
+                if check['type'] != 'agent.plugin':
+                    continue
```

Comment:
Only agent.plugin checks can be executed without the agent. How would I execute an agent.memory check from outside the agent?